### PR TITLE
[SPARK-24026][ML] Add Power Iteration Clustering to spark.ml

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

This PR adds PowerIterationClustering as a Transformer to spark.ml.  In the transform method, it calls spark.mllib's PowerIterationClustering.run() method and transforms the return value assignments (the Kmeans output of the pseudo-eigenvector) as a DataFrame (id: LongType, cluster: IntegerType).

This PR is copied and modified from https://github.com/apache/spark/pull/15770  The primary author is @wangmiao1981 

## How was this patch tested?

This PR has 2 types of tests:
* Copies of tests from spark.mllib's PIC tests
* New tests specific to the spark.ml APIs

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/clustering/PowerIterationClusteringSuite.scala [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    intercept[IllegalArgumentException] {
+      val typedData = data.select(
+        col("id"),
+        col("neighbors").cast(ArrayType(DoubleType, containsNull = false)).alias("neighbors"),
+        col("similarities")
+      )
+      model.transform(typedData)
+    }
+    intercept[IllegalArgumentException] {
+
```
Comment (wangmiao1981) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/PowerIterationClusteringSuite.scala]: remove blank line or add blank line after line 139 for consistence?

**Comment by wangmiao1981** [comment]
Take a quick look. Despite of the style failure and a minor format issue, LGTM.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      ).rdd.flatMap {
+        case Row(id: Long, nbrs: Seq[_], sims: Seq[_]) =>
+          require(nbrs.size == sims.size, s"The length of the neighbor ID list must be " +
+            s"equal to the the length of the neighbor similarity list.  Row for ID " +
+            s"$idColValue=$id has neighbor ID list of length ${nbrs.length} but similarity list " +
+            s"of length ${sims.length}.")
+          nbrs.asInstanceOf[Seq[Long]].zip(sims.asInstanceOf[Seq[Double]]).map {
+            case (nbr, similarity) => (id, nbr, similarity)
+          }
+      }
```
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala]: Add `Instrumentation`? Or I can add it in a separate PR.

Comment (jkbradley) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala]: Thanks for reminding me!  I'll do it now.

**Comment by jkbradley** [comment]
Thanks for reviewing this and for the LGTM @wangmiao1981 !  I'll merge with master now, with you as the primary author.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+  /**
+   * Param for the name of the input column for neighbors in the adjacency list representation.
+   * Default: "similarities"
+   * @group param
+   */
+  @Since("2.4.0")
+  val similaritiesCol = new Param[String](this, "similaritiesCol",
+    "Name of the input column for neighbors in the adjacency list representation.",
+    (value: String) => value.nonEmpty)
+
```
[the remark this answers is not in the document]
Comment (jkbradley) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/PowerIterationClustering.scala]: No, it's meant to be an adjacency list representation of the graph: neighborsCol has the set of neighbor vertex IDs, and similaritiesCol has the corresponding set of edge weights.

**Pull request closed without merging**